solutions/assignment-1/Command.py

```
@@ -0,0 +1,46 @@
+import re
+
+
+class Command(object):
+    """Command contains the features related to particular command. It also has validators to validate conditions while parsing."""
+
+    def __init__(self, name, usage, regex, type_allowed, no_of_arg, conflicting_commands, required_commands, overwrite):
+        self.name = name  # Command name
+        self.usage = usage  # What is the use of this command
+        self.regex = r'^' + regex + r'$'  # Regular expression which can distinguish the correct type of argument
+        self.type_allowed = type_allowed  # Type of arguments allowed for the command
```

If you have `self.regex`, do you really need this?

---

By adding a regex we can specify what type of patterns arguments should have. For example in the case of "--key" we only need to allow positive integers so providing inputs like "--key=-123" should raise an exception. If I don't use the pattern I cannot generalize the class Commands because then I am only left with options "int", "str" etc.
In other words, using regex gives me a lot of flexibility.

---

You're only using this as part of the error message. You could show the regex in place of that, right?

---

Yes, self.type_allowed is used only in the error message in case self.regex does not match with the argument. I can instead use self.regex itself to print the error message, but then I thought that the end-user may not be familiar with regex representation, so using some basic English phrase (which represents the regex) may be helpful. For ex, "--key" only allows positive integers. The error message can be represented as "123aaa is not a positive integer. A positive integer type argument is expected" or "123aaa is not of the pattern "^\+?\d+$". A "^\+?\d+$" type argument is expected". Here the first type of representation is easier to understand. To store that basic English phrase I am using self.type_allowed.